ci: change the url scheme for PRs and references

samoht opened the pull request:
The new scheme is:

- `/:user/:repo/pr/:id` for PRs
- `/:user/:repo/ref/*` for other references

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "url-scheme" git@github.com:samoht/datakit.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "url-scheme" git@github.com:samoht/datakit.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

samoht commented:
This is incomplete and requires more work.

samoht commented:
It's ready now. I am now super happy at having the ref encode and ref decode functions in different modules but I don't see where to put them otherwise.

talex5 commented on ci/src/cI_web.ml:
```diff
@@ -277,8 +280,11 @@ let routes ~logs ~ci ~server ~dashboards =
-    ("ref/:user/:repo/:id",                  fun () -> new ref_page t);
+    (":user/:repo/pr/:id",     fun () -> new pr_page t);
+    (":user/:repo/ref/:id",    fun () -> new ref_page None t);
+    (":user/:repo/branch/:id", fun () -> new ref_page (Some "heads") t);
```
Why do we have two routes to the same page?

samoht replied:
you mean `ref/$encode(heads/master)` and `branch/master` ? I can filter out the first one, but I think it is more consistent to keep it if client are just interested on raw refs (as some of the refs are neither `branch` or `tags`, e.g. `pull`)

talex5 replied:
Yes, I don't see any point in having `branch/foo` when we already have `ref/heads/foo` that does exactly the same thing. It's confusing.

samoht replied:
Because it is a useful distinction for the users? Why do we have separate tabs for tags and branches for instance?

talex5 commented on ci/src/cI_web.ml:
```diff
@@ -161,7 +161,10 @@ class pr_page t = object(self)
 
-class ref_page t = object(self)
+let unescape_ref s =
+  Uri.pct_decode s |> (fun s -> String.cuts ~sep:"/" s)
```
Agree it would be good to keep this next to `escape_ref`.

samoht commented:
Updated: 
* I've removed the `branch` and `tag` route
* I've move the `pct_decode` function close to the `pct_encode` 
* I've modified the way pct-encoding for references is done: `/` are preserved to get nicer-looking urls.
